crates/items/src/item_stack.rs

```diff
@@ -0,0 +1,237 @@
+    }
+
+    /// Sets the count for this `ItemStack`.
+    pub fn set_count(&mut self, count: u32) {
+        self.count = count
+    }
+
+    /// Splits this `ItemStack` in half, returning the
+    /// removed half. If the amount is odd, `self`
+    /// will be left with the least items. Returns the taken
+    /// half.
+    pub fn take_half(&mut self) -> ItemStack {
+        self.take((self.count as f64 / 2 as f64).ceil() as u32)
+            .unwrap()
+    }
+
+    /// Splits this `ItemStack` by removing the
+    /// specified amount. Returns the taken part.
+    pub fn take(&mut self, amount: u32) -> Result<ItemStack, ()> {
+        let count_left: i32 = self.count as i32 - amount as i32;
+        if count_left < 0 {
+            return Err(());
+        }
+        let taken = ItemStack {
+            count: amount,
+            ..self.clone()
+        };
+        self.count = count_left as u32;
+        Ok(taken)
+    }
+
+    /// Merges another `ItemStack` with this one.
+    pub fn merge_with(&mut self, other: &mut Self) -> bool {
+        if !self.has_same_type_and_damage(other) {
+            return false;
+        }
+        let new_count = (self.count + other.count).min(self.item.stack_size());
+        let amount_added = new_count - self.count;
+        self.count = new_count;
+        other.count -= amount_added;
+        true
+    }
+
+    /// Transfers up to `n` items to `other`.
+    pub fn transfer_to(&mut self, n: u32, other: &mut Self) {
+        let max_transfer = other.item.stack_size().saturating_sub(other.count);
+        let transfer = max_transfer.min(self.count).min(n);
+        self.count -= transfer;
+        other.count += transfer;
+    }
+
+    /// Damages the item by the specified amount.
+    /// If this function returns `true`, then the item is broken.
+    pub fn damage(&mut self, amount: u32) -> bool {
+        match &mut self.meta.damage {
+            Some(damage) => {
+                *damage += amount;
+                if let Some(durability) = self.item.durability() {
+                    *damage >= durability
+                } else {
+                    false
+                }
+            }
+            None => false,
+        }
+    }
+}
+
+impl Serialize for Item {
+    fn serialize<S>(&self, serializer: S) -> Result<<S as Serializer>::Ok, <S as Serializer>::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self.name())
+    }
+}
+
+struct ItemVisitor;
+
+impl<'de> Visitor<'de> for ItemVisitor {
+    type Value = Item;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "a string")
+    }
+
+    fn visit_str<E>(self, string: &str) -> Result<Self::Value, E>
+    where
+        E: Error,
+    {
+        if let Some(item) = Item::from_name(string) {
+            Ok(item)
+        } else {
+            Err(E::custom("Unknown item name."))
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Item {
```

Comment:
Use `TryFrom`

Reply:
You mean by implementing `TryFrom` and using `#[serde(try_from = "String")]` on the `Item` enum?

Reply:
Yes, I'm not sure if there are any performance complications since it is an Allocated string.

Reply:
I have implemented it and it's so compact compared to implementing `Serialize` and `Deserialize`. Didn't know of the `TryFrom` and `Into` usage in serde, thanks 😄. I will need to update the code generator for `Item`. Then I will push.

Reply:
Done!